server/src/main/java/org/elasticsearch/shutdown/PluginShutdownService.java

```diff
@@ -0,0 +1,108 @@
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.shutdown;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.apache.logging.log4j.message.ParameterizedMessage;
+import org.elasticsearch.cluster.ClusterChangedEvent;
+import org.elasticsearch.cluster.ClusterState;
+import org.elasticsearch.cluster.ClusterStateListener;
+import org.elasticsearch.cluster.metadata.NodesShutdownMetadata;
+import org.elasticsearch.cluster.metadata.SingleNodeShutdownMetadata;
+import org.elasticsearch.common.Nullable;
+import org.elasticsearch.plugins.ShutdownAwarePlugin;
+
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+/**
+ * The {@link PluginShutdownService} is used for the node shutdown infrastructure to signal to
+ * plugins that a shutdown is occurring, and to check whether it is safe to shut down.
+ */
+public class PluginShutdownService implements ClusterStateListener {
+
+    private static final Logger logger = LogManager.getLogger(PluginShutdownService.class);
+    public List<ShutdownAwarePlugin> plugins;
+
+    public PluginShutdownService(@Nullable List<ShutdownAwarePlugin> plugins) {
+        this.plugins = plugins == null ? Collections.emptyList() : plugins;
+    }
+
+    /**
+     * Return all nodes shutting down from the given cluster state
+     */
+    public static Set<String> shutdownNodes(final ClusterState clusterState) {
+        return NodesShutdownMetadata.getShutdowns(clusterState)
+            .map(NodesShutdownMetadata::getAllNodeMetadataMap)
+            .map(Map::keySet)
+            .orElse(Collections.emptySet());
+    }
+
+    /**
+     * Return all nodes shutting down with the given shutdown type from the given cluster state
+     */
+    public static Set<String> shutdownTypeNodes(final ClusterState clusterState, final SingleNodeShutdownMetadata.Type shutdownType) {
+        return NodesShutdownMetadata.getShutdowns(clusterState)
+            .map(NodesShutdownMetadata::getAllNodeMetadataMap)
+            .map(m -> m.entrySet().stream()
+                .filter(e -> e.getValue().getType() == shutdownType)
+                .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue)))
+            .map(Map::keySet)
+            .orElse(Collections.emptySet());
+    }
+
+    /**
+     * Check with registered plugins whether the shutdown is safe for the given node id and type
+     */
+    public boolean readyToShutdown(String nodeId, SingleNodeShutdownMetadata.Type shutdownType) {
```

Comment:
I wonder if it's worth having this method take an extra `ClusterState` parameter, and pass that as an extra argument to each of the `plugin.safeToShutdown` calls it makes.  It will make it easier for plugins that don't currently have their own `ClusterService` reference to implement the interface.

Reply:
We discussed adding these today, but since there's no current user, we're going to keep the cluster state out of the interface for now, and revisit it when ML (or a different plugin) has an implementation where they need these

Reply:
OK cool.  I am going to work on the ML PR soon, so can add the arguments to that if you don't have a fundamental objection.

Reply:
> so can add the arguments to that if you don't have a fundamental objection.

If possible, I think I'd prefer to keep them out of the interface. Especially for the `signalShutdown` method, if the cluster state is added it's essentially no different than a regular `ClusterStateListener` call. I think it's okay to add it to the `readyToShutdown` because that is a one-off call (not called on every cluster state change).

Would that work for you?

Reply:
Since both methods are implemented by the same class, it doesn't really help to just add the current cluster state to one of them.  I can instead add a reference to the `ClusterService` to the class that implements the interface, and then that can be used in both methods.

Reply:
That would be my preference then, as long as that isn't too distasteful of a solution for you.